compiler-rt/lib/radsan/radsan_context.cpp

```diff
@@ -0,0 +1,82 @@
+//===--- radsan_context.cpp - Realtime Sanitizer --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//===----------------------------------------------------------------------===//
+
+#include <radsan/radsan_context.h>
+
+#include <radsan/radsan_stack.h>
+
+#include <sanitizer_common/sanitizer_allocator_internal.h>
+#include <sanitizer_common/sanitizer_stacktrace.h>
+
+#include <new>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+using namespace __sanitizer;
+
+namespace detail {
+
+static pthread_key_t Key;
+static pthread_once_t KeyOnce = PTHREAD_ONCE_INIT;
+void internalFree(void *Ptr) { __sanitizer::InternalFree(Ptr); }
+
+} // namespace detail
+
+namespace radsan {
+
+Context::Context() = default;
+
+void Context::RealtimePush() { RealtimeDepth++; }
+
+void Context::RealtimePop() { RealtimeDepth--; }
+
+void Context::BypassPush() { BypassDepth++; }
+
+void Context::BypassPop() { BypassDepth--; }
+
+void Context::ExpectNotRealtime(const char *InterceptedFunctionName) {
+  if (InRealtimeContext() && !IsBypassed()) {
+    BypassPush();
+    PrintDiagnostics(InterceptedFunctionName);
+    exit(EXIT_FAILURE);
+    BypassPop();
+  }
+}
+
+bool Context::InRealtimeContext() const { return RealtimeDepth > 0; }
+
+bool Context::IsBypassed() const { return BypassDepth > 0; }
+
+void Context::PrintDiagnostics(const char *InterceptedFunctionName) {
+  fprintf(stderr,
+          "Real-time violation: intercepted call to real-time unsafe function "
+          "`%s` in real-time context! Stack trace:\n",
+          InterceptedFunctionName);
+  radsan::PrintStackTrace();
+}
+
+Context &GetContextForThisThread() {
```

Comment:
Minor: for code readability, I reckon it might be good to move the implementation of this method into a `detail::GetContextForThisThreadImpl()` method, which would live right next to the declarations of `detail::Key` and the other stuff that it uses.

Reply:
(it would save the reader from wondering what on earth `detail::Key` is used for when encountering it, for example)